src/selectors/per-thread/index.js

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+import { createSelector } from 'reselect';
+import * as UrlState from '../url-state';
+import * as ProfileData from '../../profile-logic/profile-data';
+import {
+  getThreadSelectorsPerThread,
+  type ThreadSelectorsPerThread,
+} from './thread';
+import {
+  getMarkerSelectorsPerThread,
+  type MarkerSelectorsPerThread,
+} from './markers';
+import {
+  getStackAndSampleSelectorsPerThread,
+  type StackAndSampleSelectorsPerThread,
+} from './stack-sample';
+import * as ProfileSelectors from '../profile';
+
+import type { ThreadIndex } from '../../types/profile';
+import type { Selector } from '../../types/store';
+import type { TimingsForPath } from '../../profile-logic/profile-data';
+
+/**
+ * Selectors only memoize off of the current state, and do not accept a parameter.
+ * The following selectors are duplicated for every thread that is in a profile.
+ * This way, each thread has its own memoization properties.
+ */
+export type ThreadSelectors = {|
+  ...ThreadSelectorsPerThread,
+  ...MarkerSelectorsPerThread,
+  ...StackAndSampleSelectorsPerThread,
+|};
+
+/**
+ * This is the static object store that holds the selector functions.
+ */
+const _threadSelectorsCache: { [key: ThreadIndex]: ThreadSelectors } = {};
```

Comment:
I wonder if we couldn't reuse `memoize-immutable` here (with a Map), instead of coming up with our own cache. Not for this PR that tries to not break anything, but could be done in a next PR.

Reply:
I'm not sure I see the benefit.

Reply:
The benefit is to delete our code that deals with the cache. Less code, less code to maintain. I wouldn't advise doing it if we didn't use `memoize-immutable` already.